velox/common/caching/SsdFile.cpp

```diff
@@ -753,72 +725,57 @@ void SsdFile::checkpoint(bool force) {
       // kMapMarker,
       // {fileId, offset, SSdRun} triples,
       // kEndMarker.
-      state.write(checkpointVersion().data(), sizeof(int32_t));
-      state.write(asChar(&maxRegions_), sizeof(maxRegions_));
-      state.write(asChar(&numRegions_), sizeof(numRegions_));
+      checkpointWriteFile_->append(checkpointVersion());
+      checkpointWriteFile_->append(
+          folly::IOBuf::copyBuffer(&maxRegions_, sizeof(maxRegions_)));
+      checkpointWriteFile_->append(
+          folly::IOBuf::copyBuffer(&numRegions_, sizeof(numRegions_)));
 
       // Copy the region scores before writing out for tsan.
       const auto scoresCopy = tracker_.copyScores();
-      state.write(asChar(scoresCopy.data()), maxRegions_ * sizeof(uint64_t));
+      checkpointWriteFile_->append(folly::IOBuf::copyBuffer(
+          scoresCopy.data(), maxRegions_ * sizeof(double)));
       std::unordered_set<uint64_t> fileNums;
       for (const auto& entry : entries_) {
         const auto fileNum = entry.first.fileNum.id();
```

Comment:
You could have a method to estimate the entry size and keep accumulating to the buffer until reach the limit, and then flush once.